### PR TITLE
Config refactor

### PR DESCRIPTION
**Pull request by erikrose**

### CODE REVIEW

File: dxr/config.py [hunk 1]
```diff
@@ -16,215 +24,270 @@
...
+                Optional('temp_folder', default='/tmp/dxr-temp'): AbsPath,
+                Optional('default_tree', default=None): basestring,
+                Optional('disabled_plugins', default=plugin_list('')): Plugins,
+                Optional('enabled_plugins', default=plugin_list('*')): Plugins,
+                Optional('generated_date',
+                         default=datetime.utcnow()
+                                         .strftime("%a, %d %b %Y %H:%M:%S +0000")):
+                    basestring,
+                Optional('log_folder', default=None): AbsPath,
+                Optional('workers', default=cpu_count()):
```
Comment (abbeyj) [inline dxr/config.py]: The docs say that multiprocessing.cpu_count() can raise NotImplementedError.  I think we should catch that and use the value 1 instead.

File: dxr/config.py [hunk 2]
```diff
@@ -16,215 +24,270 @@
...
+                    Use(int, error='"workers" must be an integer >= 1.'),
```
Comment (abbeyj) [inline dxr/config.py]: The message should say that 0 is allowed too, right?  What happens if the user does something silly like pass "-5" for this?

File: dxr/indexers.py [hunk 3]
```diff
@@ -9,7 +9,20 @@
 from os.path import join
 
 
-class TreeToIndex(object):
+class PluginConfig(object):
```
Comment (abbeyj) [inline dxr/indexers.py]: Maybe say that this is a mixin?

File: dxr/plugins/buglink/tests/test_links/dxr.config.in [hunk 4]
```diff
@@ -9,5 +9,7 @@ es_alias            = dxr_test_{format}_{tree}
 source_folder       = PWD/code
 object_folder       = PWD/code
 build_command       = /bin/true
-plugin_buglink_name = Fred's Bugzilla
-plugin_buglink_url = http://bugs.com/%s
\ No newline at end of file
+
+    [[buglink]]
+    name = Fred's Bugzilla
+    url = http://bugs.com/%s
```
Comment (abbeyj) [inline dxr/plugins/buglink/tests/test_links/dxr.config.in]: Is it worth changing this to "bugs.example.com"?

**Comment by erikrose** [comment]
All good stuff, abbeyj. I'll make the fixes.

File: dxr/config.py (hunk 2 above)
Comment (erikrose) [inline dxr/config.py]: Then they get the late failure they deserve. ;-) Okay, I'll patch this up as well.

File: dxr/plugins/buglink/tests/test_links/dxr.config.in (hunk 4 above)
Comment (erikrose) [inline dxr/plugins/buglink/tests/test_links/dxr.config.in]: Since you thought to ask, yes. :-)

**Pull request merged**